Add support for other HTTP verbs for aggregation aka `UpstreamHttpMethod` enhancements

crb02005 opened the pull request:
This changes the FileAggregateRoute to allow other verbs. 
It is for consumers not following traditional REST. Several APIs do not follow the REST guidelines and may use POST for getting data. In cases like these being able to specify a verb is helpful.

crb02005 committed:
Add support for other verbs for aggregation, for consumers not following traditional REST

raman-m commented:
Hi Carl!
Thanks for your interest in Ocelot!

I will review...

Could you [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) please? So, your **develop** branch is outdated!

raman-m commented on src/Ocelot/Configuration/File/FileAggregateRoute.cs:
```diff
@@ -10,10 +10,7 @@ public class FileAggregateRoute : IRoute
-        // Only supports GET..are you crazy!! POST, PUT WOULD BE CRAZY!! :)
-        public List<string> UpstreamHttpMethod => new() { "Get" };
-
+        public List<string> UpstreamHttpMethod { get; private set; } = new();
```
`private set` means you can assign the value by constructor only.
In this case default constructor creates the list.
No much sense to make setter private:
- another properties have public setter
- we can unit test this property

raman-m committed:
code review

raman-m commented on src/Ocelot/Configuration/Creator/AggregatesCreator.cs:
```diff
@@ -26,6 +25,12 @@ public List<Route> Create(FileConfiguration fileConfiguration, List<Route> route
 
         private Route SetUpAggregateRoute(IEnumerable<Route> routes, FileAggregateRoute aggregateRoute, FileGlobalConfiguration globalConfiguration)
         {
+            if (!aggregateRoute.UpstreamHttpMethod.Any())
+            {
+                // Default method to Get for standard use case
+                aggregateRoute.UpstreamHttpMethod.Add(HttpMethod.Get.ToString());
+            }
+
```
We have to unit test this code.

raman-m replied:
And what's the goal of having static field, Raynald?
I have another alternative which is static too &rarr; `HttpMethods.Get` in the `Microsoft.AspNetCore.Http` namespace 😉
This is fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1389/commits/02a3d991c4eb4d8ae04f8ff364046bd7a35aceb3 ✔️

raman-m replied:
@RaynaldM, I've implemented your concept of a single global setting. The commit is https://github.com/ThreeMammals/Ocelot/pull/1389/commits/98083a1e3cb503723e5db8e8ea0c8b564dced01b. However, this setting can only be overridden in the C# code; there's no corresponding JSON setting in `ocelot.json`.
Also, I chose to add the global static property to the `FileAggregateRoute` class rather than the `AggregatesCreator` because it appears to be the most appropriate place for its definition.

raman-m replied:
Unit tests commit is https://github.com/ThreeMammals/Ocelot/pull/1389/commits/7e071939f553d0054179740f4d4b2929e64942ca ✔️

raman-m reviewed (changes requested):
Everything is OK.
But we have to write a few acceptance tests which will cover added logic.

Could you write at least one acceptance test please?

raman-m committed:
Add required namespace

raman-m commented:
The [feature branch](/crb02005/Ocelot/tree/user/carl-aggregate-methods) has been [rebased](#event-10157666704) onto [ThreeMammals:develop](../tree/develop)!
Welcome to code review!

---

I see that **develop** branch in [your fork](/crb02005/Ocelot/branches) is too old!
Could you [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) please? So, **develop** branch will be updated with top commits!
Could you [add me as collaborator](/crb02005/Ocelot/settings/access) to your forked repo please?

raman-m committed:
EOL: src/Ocelot/Configuration/Creator/AggregatesCreator.cs

raman-m committed:
EOL: test/Ocelot.AcceptanceTests/AggregateTests.cs

raman-m committed:
@RaynaldM loves static fields

raman-m committed:
Refactor `FileAggregateRoute`:
- sort lines and props
- move initialization to def ctor

raman-m committed:
Refactor `AggregatesCreatorTests`:
- file-scoped namespace
- IDE1006 (Naming rule violation)
- remove BDDfy
- AAA-pattern
- add traits

raman-m committed:
Refactor AggregatesCreatorTests

raman-m committed:
@RaynaldM is likely to appreciate this static property even more

raman-m committed:
Add unit tests

raman-m committed:
Move unit tests

raman-m commented on docs/features/requestaggregation.rst:
```diff
@@ -218,7 +218,7 @@ Gotchas
 
 * You cannot use Routes with specific **RequestIdKeys** as this would be crazy complicated to track.
-* Aggregation only supports the ``GET`` HTTP verb.
+* Aggregation supports the ``GET`` HTTP verb for pure REST, it supports other verbs for APIs which do not fully follow REST.
```
Update docs with info about global static `DefaultHttpMethod`

raman-m commented on test/Ocelot.AcceptanceTests/AggregateTests.cs:
```diff
@@ -653,6 +653,12 @@ public void Should_return_response_200_with_copied_form_sent_on_multiple_service
                 .BDDfy();
         }
 
+        [Fact]
+        [Trait("Feat", "1389")]
+        public void TODO()
+        {
+        }
```
Write acceptance tests

raman-m reviewed (commented):
### Ready for Code Review❕

raman-m commented:
@RaynaldM @ggnaegi Greetings team! **Your assistance and recommendation are required.**
Could you please review the implementation once more?
I have introduced a global static property for the default `GET` method in the creator's class, but it appears that the feature will remain incomplete without a genuine JSON option. Kindly refer to [my comment](https://github.com/ThreeMammals/Ocelot/pull/1389#discussion_r1831483227) for Ray's previous code review.

@crb02005 Hi Carl! Are you online?
What are your thoughts on a global JSON setting for the HTTP methods in aggregated routes? This feature would allow us to set desired HTTP verbs at the route level, but what about having one global JSON option?

raman-m committed:
Merge 'ThreeMammals:develop' into 'crb02005:user/carl-aggregate-methods'

raman-m committed:
Update docs
